<h5>
Inheritance mode values are candidate labels based on genotype assignments.
</h5>

<p>
This case is labeled as proband-only. Since no other family members are in the analysis, most of the traditional inheritance modes can not be assigned.
</p>

<p>
<b>Inheritance modes that can not be assigned on this case include:</b>
<ul>
<li> de novo calls
<li> Dominant (Maternal) vs Dominant (Paternal) vs Recessive
<li> Loss of Heterozygosity
</ul>
</p>

<p>
Variants on the X or Y chromosome are classified <b>X-linked</b> or <b>Y-linked</b>.
</p>

<p>
<b>Low relevance</b>:
<ul>
<li> Variants with missing calls (./.) or calls that mismatch the sex (e.g. 0/1 on chrX for a male) are not assigned inheritance mode values.
<li> Variants with genotypes that do not meet any of the high relevance criteria above are not assigned inheritance mode labels.
</p>

<p><b>Ambiguous</b>: </p>

<p>
  If a sample was submitted to CGAP with the sex indicated as "U" or "Unknown", then autosome inheritance modes and genotypes
  will still be calculated normally but sex chromosome inheritance mode and genotype cannot be properly determined. These will
  be marked ambiguous.
</p>
